### ISSUE TITLE
Error-deleting-CloudWatch-Rule

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#1479]
_This issue was originally opened by @mhmdio as hashicorp/terraform#15863. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version
Terraform v0.9.11

### Terraform Configuration Files
```hcl
resource "aws_cloudwatch_event_rule" "cloudwatch_to_s3" {
  name                = "cloudwatch-to-s3"
  description         = "copy cloudwatch VPC flow logs to s3"
  schedule_expression = "cron(0 0 * * ? *)"
}

resource "aws_cloudwatch_event_target" "cloudwatch_to_s3" {
  rule      = "${aws_cloudwatch_event_rule.cloudwatch_to_s3.name}"
  target_id = "cloudwatch_to_s3"
  arn       = "${aws_lambda_function.cloudwatch_to_s3.arn}"
}
```

### Debug Output
https://gist.github.com/mhmdio/e2ef142eda0c0d667583b35f5716815e

### Expected Behavior
Terraform should remove CW Event targets then remove the rule.

### Actual Behavior
Terraform try to remove the CW Event rule before deleting CW Event targets.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `create a module from gist resources`
2. `terraform get`
3. `terraform apply`

**Comment by bflad** [comment]
This is likely due to the fact that many AWS services including CloudWatch are eventually consistent. Given the configuration above, Terraform should have ordered the operations properly however CloudWatch likely had not fully propagated the target deletion internally before Terraform sent the rule deletion request. In many places we handle this by adding retry code around these operations.

**Comment by AleksandarTokarev** [comment]
Any solution to this problem yet?

**Comment by jgspratt** [comment]
What workarounds are available for this problem?

**Comment by bostrowski13** [comment]
2 years since the issue was reported with no resolution.  when can this be fixed, or a workaround provided?

**Comment by pprieto** [comment]
Same issue here

**Comment by sirocode** [comment]
Any news? This annoying bug breaks our pipelines regularly.

**Comment by bflad** [comment]
Potential fix submitted: https://github.com/terraform-providers/terraform-provider-aws/pull/11475

Please see the pull request description for additional information.

**Comment by bflad** [comment]
The eventual consistency retries and `RemoveTargets` API call `FailedEntry` handling have been merged and will release with version 2.45.0 of the Terraform AWS Provider, Thursday this week.

If you are still encountering issues after the release of version 2.45.0, please open a new [bug report](https://github.com/terraform-providers/terraform-provider-aws/issues/new?template=Bug_Report.md) following the issue template and we will take a fresh look. 👍
